src/lib.rs

```diff
@@ -299,92 +322,208 @@ pub fn write_b(cw: &mut ContextWriter, fi: &FrameInvariants, fs: &mut FrameState
+        let d = sse_wxh(&fs.input.planes[0].slice(&po), &fs.rec.planes[0].slice(&po),
+                        w as usize, h as usize);
+        let r = ((cw.w.tell_frac() - tell) as f64)/8.0;
+
+        let rd = (d as f64) + lambda*r;
+        if rd < best_rd {
+            best_rd = rd;
+            best_mode = mode;
+        }
+
+        cw.rollback(checkpoint.clone());
+    }
+
+    assert!(best_rd as i64 >= 0);
+
+    let rdo_output = RDOOutput { rd_cost: best_rd as u64,
+                                pred_mode: best_mode};
+    rdo_output
+}
+
+fn write_sb(fi: &FrameInvariants, fs: &mut FrameState, cw: &mut ContextWriter,
+            bsize: BlockSize, bo: &BlockOffset) {
+
+    if bo.x >= cw.bc.cols || bo.y >= cw.bc.rows {
+        return;
+    }
 
-    // Update partition context
-    let subsize = get_subsize(bsize, PartitionType::PARTITION_NONE);
+    // TODO(anyone): Until we have RDO-based block size decision,
+    // split all the way down to 8x8 blocks, then do rdo_mode_decision() for each 8x8 block.
+    // FIXIME(anyone): Enable partition into 4x4 blocks.
+    let mut partition = PartitionType::PARTITION_SPLIT;
 
-    cw.bc.update_partition_context(&po, subsize, bsize);
+    if bsize <= BlockSize::BLOCK_8X8 {
+        partition = PartitionType::PARTITION_NONE;
+    }
+
+    assert!(mi_size_wide[bsize as usize] == mi_size_high[bsize as usize]);
+
+    assert!(PartitionType::PARTITION_NONE <= partition &&
+            partition < PartitionType::PARTITION_INVALID);
+
+    let bs = mi_size_wide[bsize as usize];
+    let hbs = bs >> 1; // Half the block size in blocks
+    let subsize = get_subsize(bsize, partition);
+
+    if bsize >= BlockSize::BLOCK_8X8 {
+        cw.write_partition(bo, partition, bsize);
+    }
+
+    match partition {
+        PartitionType::PARTITION_NONE => {
+            // TODO(anyone): Until we have RDO-based block size decision,
+            // call rdo_mode_decision() for a partition.
+            let rdo_none = rdo_mode_decision(fi, fs, cw, bsize, bo);
+            // FIXME(anyone): Instead of calling set_mode() in write_b() for each 4x4tx position,
+            // it would be better to call set_mode() for each MI block position here.
+            cw.bc.set_mode(&bo, rdo_none.pred_mode);
+
+            write_b(fi, fs, cw, rdo_none.pred_mode, bsize, bo);
+        },
+        PartitionType::PARTITION_SPLIT => {
+            assert!(subsize != BlockSize::BLOCK_INVALID);
+            write_sb(fi, fs, cw, subsize, bo);
+            write_sb(fi, fs, cw, subsize, &BlockOffset{x: bo.x + hbs as usize, y: bo.y});
+            write_sb(fi, fs, cw, subsize, &BlockOffset{x: bo.x, y: bo.y + hbs as usize});
+            write_sb(fi, fs, cw, subsize, &BlockOffset{x: bo.x + hbs as usize, y: bo.y + hbs as usize});
+        },
+        _ => { assert!(false); },
+    }
+
+    if bsize >= BlockSize::BLOCK_8X8 &&
+       (bsize == BlockSize::BLOCK_8X8 || partition != PartitionType::PARTITION_SPLIT) {
```

Comment:
isn't equivalent to `bsize == BlockSize::BLOCK_8X8 || (bsize > BlockSize::BLOCK_8X8 && partition != PartitionType::PARTITION_SPLIT)` ?

Reply:
You are right in its absolute meaning if the meaning there is not counted.
First, the partition info is not coded for BLOCK_4X4 since it is the smallest block, then update_partition() is only considered for BLOCK_8x8 or larger, hence the first condition check it.
Another reason is, I try to follow exactly the same condition as in libaom for easier migration from existing libaom codebase.